src/hotspot/share/classfile/stringTable.hpp

```diff
@@ -104,14 +105,42 @@ class StringTable : public CHeapObj<mtSymbol>{
   }
 
   // Sharing
+#if INCLUDE_CDS_JAVA_HEAP
+  static inline oop read_string_from_compact_hashtable(address base_address, u4 index);
+
+private:
+  static bool _two_dimensional_shared_strings_array;
+  static OopHandle _shared_strings_array;
+  static int _shared_strings_array_root_index;
+
+  // All the shared strings are referenced through _shared_strings_array to keep them alive.
+  // Each shared string is stored as a 32-bit index in ::_shared_table. The index
+  // is interpreted in two ways:
+  //
+  // [1] _two_dimensional_shared_strings_array = false: _shared_strings_array is an Object[].
+  //     Each shared string is stored as _shared_strings_array[index]
+  //
+  // [2] _two_dimensional_shared_strings_array = true: _shared_strings_array is an Object[][]
+  //     This happens when there are too many elements in the shared table. We store them
+  //     using two levels of objArrays, such that none of the arrays are too big for
+  //     ArchiveHeapWriter::is_too_large_to_archive(). In this case, the index is splited into two
+  //     parts using the follow bit-opts. Each the shared string is stored as
+  //     _shared_strings_array[upper][lower]
```

Comment:
upper/lower don't match use of primary/secondary. I'm still unclear how the 32-bit index is split in the two dimensional case.

Reply:
I reworded the comments and use the same terminology as in the actual code (primary_index, secondary_index).